docs/design-documents/20240430-schema-support.md

```diff
@@ -0,0 +1,421 @@
+    - [Internal schema format](#internal-schema-format)
+      - [Option 1: Avro](#option-1-avro)
+      - [Option 2: Protobuf schema](#option-2-protobuf-schema)
+    - [Chosen option](#chosen-option-1)
+  - [Schema service interface](#schema-service-interface)
+    - [Option 1: Stream of commands and responses](#option-1-stream-of-commands-and-responses)
+    - [Option 2: Exposing a gRPC service in Conduit](#option-2-exposing-a-grpc-service-in-conduit)
+    - [Chosen option](#chosen-option-2)
+- [Required changes](#required-changes)
+  - [Conduit](#conduit)
+  - [Conduit Commons](#conduit-commons)
+  - [Connector SDK](#connector-sdk)
+  - [Processor SDK](#processor-sdk)
+- [Summary](#summary)
+- [Other considerations](#other-considerations)
+
+## The goal
+
+The goal of schema support is to make the information about data types available
+in a pipeline. This is needed to cover a few use cases, such as automatically
+creating a destination collection or propagating schema changes.
+
+## Requirements
+
+1. Records **should not** carry the whole schema.
+
+   Reason: If a record would carry the whole schema, that might increase a
+   record's size significantly.
+2. Sources and destinations need to be able to work with multiple schemas.
+
+   Reason: Multiple collections support.
+3. A schema should be accessible across pipelines and Conduit instances.
+
+   Reason: Pipelines might work together to stream data from A to B, through an
+   intermediary C. For example, PostgreSQL to Kafka, and then Kafka to
+   Snowflake. Also, in the future, pipelines might run in isolation on different
+   instances. Additionally, Conduit may be used so that there's one pipeline per
+   instance. In such cases, we'd need this feature.
+4. It should be possible for a schema to evolve.
+5. A source connector should be able to register a schema.
+6. A destination connector should be able to fetch a specific schema.
+7. A destination connector needs a way to know that a schema changed.
+8. The Connector SDK should provide an API to work with the schemas.
+9. The Connector SDK should cache the schemas.
+
+   Reason: Standalone connectors and Conduit communicate via gRPC. To avoid the
+   cost of repeatedly fetching or creating the same schema many times (
+   especially over gRPC), schemas should be cached by the SDK.
+10. Schema auto-generation
+
+    Reason: Schema auto-generation is a useful feature when working with data
+    sources that have structured data encoded into a binary format (e.g. a JSON
+    file). In such cases, it's useful to have the schema auto-generated as a
+    starting point for further processing and/or writing into a destination.
+
+## Schema structure
+
+A schema consists of following:
+
+- reference: a string that uniquely identifies a schema in Conduit
+- list of fields, where each field is described with following:
+  - name
+  - type
+  - optional (boolean value)
+  - default value
+
+The following types are supported:
```

Comment:
The supported types really depend on the schema format, I'd remove this part to not mislead anyone. Maybe instead mention that multiple schema types will be supported, but we will start with Avro, as mentioned in the `Schema format` chapter.

Reply:
You're right about that. The point of this section was to mention what we consider are must-haves, i.e. requirements for a schema whatever the schema might be.

Reply:
@hariso it's possible that all data serialization systems have those types,  but to make it super clear, I would mention that the types will depend on the system used. Maybe adding a link to avro's for reference since we're starting with that one https://avro.apache.org/docs/1.11.1/specification/#schema-declaration.

In a second thought, maybe we could mention those would be the types we'll come up with once we bring our own schema structure. In other words, the whole Schema structure will be **how it'll look like**, but https://avro.apache.org/docs/1.11.1/specification/#schema-declaration is what we're starting with.

Reply:
@raulb You're right about that, the types will depend on the actual schema format being used. I can add that to the doc.

Here's a bit of context: I wrote the requirements prior to us choosing the schema format, so we know what are the minimum requirements for a schema. For example, Lyubo asked for a timestamp type (for SQL connectors).